Mathlib/Analysis/CstarAlgebra/HilbertCstarModule.lean

```diff
@@ -0,0 +1,286 @@
+/-
+Copyright (c) 2024 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import Mathlib.Analysis.CstarAlgebra.ContinuousFunctionalCalculus.Order
+
+/-!
+# Hilbert C⋆-modules
+
+A Hilbert C⋆-module is a complex module `E` together with a right `A`-module structure, where `A`
+is a C⋆-algebra, and with an "inner product" that takes values in `A`. This inner
+product satisfies the Cauchy-Schwarz inequality, and induces a norm that makes `E` a normed
+vector space.
+
+## Main declarations
+
++ `HilbertCstarModule`: The class containing the Hilbert C⋆-module structure
++ `HilbertCstarModule.normedSpaceCore`: The proof that a Hilbert C⋆-module is a normed vector
+  space. This can be used with `NormedAddCommGroup.ofCore` and `NormedSpace.ofCore` to create
+  the relevant instances on a type of interest.
++ `HilbertCstarModule.inner_mul_inner_swap_le`: The statement that
+  `⟪y, x⟫ * ⟪x, y⟫ ≤ ‖x‖ ^ 2 • ⟪y, y⟫`, which can be viewed as a version of the Cauchy-Schwarz
+  inequality for Hilbert C⋆-modules.
++ `HilbertCstarModule.norm_inner_le`, which states that `‖⟪x, y⟫‖ ≤ ‖x‖ * ‖y‖`, i.e. the
+  Cauchy-Schwarz inequality.
+
+## Implementation notes
+
+The class `HilbertCstarModule A E` requires `E` to already have a `Norm E` instance on it, but
+no other norm-related instances. We then include the fact that this norm agrees with the norm
+induced by the inner product among the axioms of the class. Furthermore, instead of registering
+`NormedAddCommGroup E` and `NormedSpace ℂ E` instances (which might already be present on the type,
+and which would send the type class search algorithm on a chase for `A`), we provide a
+`NormedSpace.Core` structure which enables downstream users of the class to easily register
+these instances themselves on a particular type.
+
+## References
+
++ Erin Wittlich. *Formalizing Hilbert Modules in C⋆-algebras with the Lean Proof Assistant*,
+  December 2022. Master's thesis, Southern Illinois University Edwardsville.
+-/
+
+open scoped ComplexOrder RightActions
+
+/-- A *Hilbert C⋆-module* is a complex module `E` endowed with a right `A`-module structure
+(where `A` is typically a C⋆-algebra) and an inner product `⟪x, y⟫_A` which satisfies the
+following properties. -/
+class HilbertCstarModule (A : outParam <| Type*) (E : Type*) [NonUnitalSemiring A] [StarRing A]
+    [Module ℂ A] [AddCommGroup E] [Module ℂ E] [PartialOrder A] [SMul Aᵐᵒᵖ E] [Norm A] [Norm E]
```

Comment:
I already got quite far with this: I have a complete proof that matrices with entries in a C*-algebra is a C*-algebra that I'll probably PR tonight, and I have not felt any pain related to this at all. Actually, in a previous version I tried to avoid the right `A`-modules and it was horrible: you do need right multiplication somewhere, and I think this is much less painful than the alternatives, like having the matrices act on the right on row vectors.

Reply:
As for getting to applications where we really need to work with `NonUnitalModule` in a nontrivial way: is it urgent enough to resort to hacks instead of just biting the bullet and actually writing `NonUnitalModule`?